diff --git a/include/openmc/particle_data.h b/include/openmc/particle_data.h
index ec393845f..ce8d7036a 100644
--- a/include/openmc/particle_data.h
+++ b/include/openmc/particle_data.h
@@ -423,6 +423,8 @@ private:
 #endif
 };
 
+
+
 //============================================================================
 //! Defines how particle data is laid out in memory
 //============================================================================
diff --git a/include/openmc/random_ray/random_ray.h b/include/openmc/random_ray/random_ray.h
index b8d1bc228..04944f20b 100644
--- a/include/openmc/random_ray/random_ray.h
+++ b/include/openmc/random_ray/random_ray.h
@@ -22,6 +22,11 @@ struct RayBufferContainer {
   // int receiving_rank;
   bool is_active;
   uint64_t ray_id; 
+  #ifdef OPENMC_DAGMC_ENABLED
+  int n_handles; // will populate as history_.size()
+  moab::EntityHandle handles[5];
+  Direction last_dir; // Can read/assign from rays directly (if DAGMC macro is defined)
+  #endif
 };
 
 // Container for MPI exchange
@@ -31,7 +36,12 @@ struct RayExchangeData {
   double distance_travelled;
   int surface;
   bool is_active;
-  uint64_t ray_id; 
+  uint64_t ray_id;
+  #ifdef OPENMC_DAGMC_ENABLED
+  int n_handles; // will populate as history_.size()
+  moab::EntityHandle handles[5];
+  Direction last_dir; // Can read/assign from rays directly (if DAGMC macro is defined)
+  #endif
 };
 
 // Forward declare
diff --git a/src/random_ray/random_ray.cpp b/src/random_ray/random_ray.cpp
index e2656d933..208459bfc 100644
--- a/src/random_ray/random_ray.cpp
+++ b/src/random_ray/random_ray.cpp
@@ -1012,6 +1012,13 @@ void RandomRay::restart_ray(FlatSourceDomain* domain, RayExchangeData& data, vec
     }
   }
 
+  #ifdef OPENMC_DAGMC_ENABLED 
+  last_dir_ = data.last_dir;
+  for (int i = data.n_handle-1; i >= 0; i--) {
+    history_.add_entity(data.handles[i]);
+  }
+  #endif
+
   // printf("RANK %d: Restart ray %ld, position: %f, %f, %f, angular flux: %f, distance travlled: %f, rank %d\n", mpi::rank, id(), r().x, r().y, r().z, angular_flux_[0], distance_travelled_, mpi::rank);
 
 }
@@ -1180,6 +1187,22 @@ void RandomRay::pack_ray_for_buffer(double distance_buffer, Position position_bu
 //  exchange_data_.receiving_rank = owner_rank_;
 
 //  is_buffered_ = true; 
+
+  #ifdef OPENMC_DAGMC_ENABLED
+  exchange_data_.last_dir_ = last_dir_;
+  exchange_data_.n_handles = history_.size();
+  if (history_.size() > 5) {
+    fatal_error(fmt::fmt("Ray history size {} exceeds allocated handle "
+                            "buffer size 5 for ray {}.",
+                            history_.size(), id()));
+  }
+  for (int i = 0; i < exchange_data_.n_handles; i++) {
+    moab::EntityHandle handle;
+    history_.get_last_intersection(handle);
+    exchange_data_.handles[i] = handle;
+    history_.rollback_last_intersection();
+  }
+  #endif
 }
 
 int RandomRay::get_energy_groups() {
diff --git a/src/random_ray/ray_bank.cpp b/src/random_ray/ray_bank.cpp
index 214ceb24b..60ae3585e 100644
--- a/src/random_ray/ray_bank.cpp
+++ b/src/random_ray/ray_bank.cpp
@@ -167,6 +167,14 @@ void RayBank::communicate_rays(){
           // }
         }
         // simulation::time_test.stop();
+        #ifdef OPENMC_DAGMC_ENABLED
+          exchange_data.n_handles = rays[i].n_handles;
+          for (int j = 0; j < rays[i].n_handles; j++) {
+            exchange_data.handles[j] = rays[i].handles[j];
+          }
+          exchange_data.last_dir = rays[i].last_dir;
+        #endif
+
       }
 
       MPI_Isend(&ray_data[vector_send_idx], num_rays_sending * sizeof(RayExchangeData), MPI_BYTE, receiving_rank, 1, mpi::intracomm, &requests[req_idx]);
